Separate downloaded files from tracked ones

obrakmann opened the pull request:
This change puts all downloaded files in a separate directory, so that tracked files don't get cached anymore.

Side effects are that building Windows installers on Linux works again, the build scripts produce less distracting output, and [travis now uses caching](http://docs.travis-ci.com/user/caching), too.

obrakmann committed:
Keep downloaded files in a separate directory

As a side effect, this re-enables building the Windows installer on Linux hosts.

obrakmann committed:
Make packaging less noisy, + random cleanups

obrakmann committed:
Adjust travis-ci caching

1. Removes the APT caching, since it's not available on the docker infrastructure.
2. Enables caching for the dependency download directory.

Mailaender commented on Makefile:
```diff
@@ -286,20 +286,19 @@ all: cli-dependencies core tools
 	@-$(RM_F) *.exe *.dll ./OpenRA*/*.dll ./OpenRA*/*.mdb *.mdb mods/**/*.dll mods/**/*.mdb *.resources
 	@-$(RM_RF) ./*/bin ./*/obj
-	@-$(RM_F) ./thirdparty/*.dll
-	@-$(RM_RF) ./thirdparty/windows
```
We could use

```bash
@-$(RM_RF) ./thirdparty/download/windows
```

here too.

Mailaender replied:
Nevermind. Recursion.

Mailaender commented:
This will make us resistant against nuget.org downtimes. :+1:

Mailaender commented:
Caching works:
* https://travis-ci.org/OpenRA/OpenRA/builds/61022080
* https://travis-ci.org/OpenRA/OpenRA/builds/61023059

Mailaender commented:
Also AppVeyor does still cache in consecutive runs:
* https://ci.appveyor.com/project/OpenRA/openra/build/1.0.2811
* https://ci.appveyor.com/project/OpenRA/openra/build/1.0.2812

Mailaender commented:
:white_check_mark:

pchote commented:
The `SDL2-CS.dll.config` is no longer copied into the game dir by `make dependencies`.

pchote commented:
Could you please rebase past #8082 and adjust the config script to download the OSX deps into the new download dir?

obrakmann commented:
Rebased, updated and fixed the missing SDL2-CS.dll.config.

I moved the OS X dependency fetching into its own script. I also changed the Makefile targets around some.

Just `make` will only rebuild the game and not fetch dependencies. `make all` will do everything, as the name implies: fetch and configure dependencies, and build all game components. `make dependencies` will just fetch and configure the dependencies and do the right thing regardless whether it's run on Linux or OS X (pchote, please test that).

Phrohdoh commented:
`configure-native-deps.sh` is now linux-specific.

pchote commented on Makefile:
```diff
@@ -271,38 +279,47 @@ $(foreach prog,$(PROGRAMS),$(eval $(call BUILD_ASSEMBLY,$(prog))))
 tools: editor gamemonitor
 
-package: cli-dependencies core editor gamemonitor docs version
+package: all-dependencies core tools docs version
```
This will pull down the linux dependencies when running the packaging scripts, which we don't want (each platform's packaging script should handle those themselves).

pchote replied:
It looks like we dont end up packaging the other platform's deps (at least OSX doesn't), but it's not clear to me why... so I guess this is ok.

obrakmann replied:
First of all, we don't have any linux-exclusive bundled dependencies anymore. And the reason why this works is that the os-specific dependencies don't get copied into `packaging/built` by `make install`. OS X and Linux build a second tree based on `packaging/built` to create packages from, so their dependencies never end up in the "template", and NSIS cherry-picks the file it needs from `packaging/built`.

pchote commented on Makefile:
```diff
@@ -271,38 +279,47 @@ $(foreach prog,$(PROGRAMS),$(eval $(call BUILD_ASSEMBLY,$(prog))))
+linux-dependencies: cli-dependencies linux-native-dependencies
+
+linux-native-dependencies:
+	@./thirdparty/configure-native-deps.sh
```
Maybe rename this to include `linux`?

pchote commented on Makefile:
```diff
@@ -271,38 +279,47 @@ $(foreach prog,$(PROGRAMS),$(eval $(call BUILD_ASSEMBLY,$(prog))))
-	@./thirdparty/configure-native-deps.sh
+dependencies: $(os-dependencies)
+
+all-dependencies: cli-dependencies windows-dependencies osx-dependencies
```
We want to make sure we don't end up packaging linux/windows deps in the OSX app bundle, and so on.

pchote commented:
Everything appears to work as expected under OS X.

obrakmann commented:
Here's a [Windows installer](https://seacloud.cc/d/5d1adbebce/files/?p=/openra/OpenRA-test-20150507.exe) built from this branch so that the Windows people can check out the package contents.

obrakmann commented:
Updated

penev92 commented:
Oh this is going to conflict so bad with #8101  :S

obrakmann commented:
Rebased.

Mailaender commented:
Time to get this merged. I take your comments as :+1: .

Mailaender commented on Makefile:
```diff
@@ -271,38 +279,47 @@ $(foreach prog,$(PROGRAMS),$(eval $(call BUILD_ASSEMBLY,$(prog))))
+linux-dependencies: cli-dependencies linux-native-dependencies
+
+linux-native-dependencies:
+	@./thirdparty/configure-linux-native-deps.sh
```
You can append this to the target above similar to `osx-dependencies` for consistency.

Mailaender replied:
Or maybe not. Someday we may have system dependencies for Linux.
